fuel-p2p/src/discovery.rs

```diff
@@ -0,0 +1,498 @@
+use self::mdns::MdnsWrapper;
+use futures::FutureExt;
+use futures_timer::Delay;
+use libp2p::{
+    core::{
+        connection::{ConnectionId, ListenerId},
+        ConnectedPoint,
+    },
+    kad::{handler::KademliaHandlerProto, store::MemoryStore, Kademlia, KademliaEvent, QueryId},
+    mdns::MdnsEvent,
+    swarm::{
+        DialError, IntoProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction, PollParameters,
+        ProtocolsHandler,
+    },
+    Multiaddr, PeerId,
+};
+use std::{
+    collections::{HashMap, HashSet, VecDeque},
+    io,
+    task::{Context, Poll},
+    time::Duration,
+};
+mod discovery_config;
+mod mdns;
+pub use discovery_config::DiscoveryConfig;
+
+const SIXTY_SECONDS: Duration = Duration::from_secs(60);
+
+/// Event generated by the `DiscoveryBehaviour`.
+#[derive(Debug)]
+pub enum DiscoveryEvent {
+    /// Notify the swarm of a Kademlia event
+    KademliaEvent(KademliaEvent),
+
+    /// Notify the swarm of a connected peer
+    Connected(PeerId),
+
+    /// Notify the swarm of a disconnected peer
+    Disconnected(PeerId),
+}
+
+/// NetworkBehavior for discovery of nodes
+pub struct DiscoveryBehaviour {
+    /// Store the local peer id
+    local_peer_id: PeerId,
+
+    /// Store Peer IDs of currently connected peers
+    connected_peers: HashSet<PeerId>,
+
+    /// Store all peer addresses of currently connected peers
+    connected_peer_addresses: HashMap<PeerId, Vec<Multiaddr>>,
+
+    /// Predefined list of nodes and their addresses
+    predefined_nodes: Vec<(PeerId, Multiaddr)>,
+
+    /// Events to report to the swarm
+    events: VecDeque<DiscoveryEvent>,
+
+    /// For discovery on local network, optionally available
+    mdns: MdnsWrapper,
+
+    /// Kademlia with MemoryStore
+    kademlia: Kademlia<MemoryStore>,
+
+    /// If enabled, the Stream that will fire after the delay expires,
+    /// starting new random walk
+    next_kad_random_walk: Option<Delay>,
+
+    /// The Duration for the next random walk, after the current one ends
+    duration_to_next_kad: Duration,
```

Comment:
did you mean timeout or this is a limit for each random walk? if the latter, then why not simply name it `kad_random_walk_duration`?

Reply:
it's not a limit, it is what the name suggests: duration to next kademlia, once the current kademlia walk ends, the duration of the incoming one